### PR TITLE
database connector updates to support some joining database connector use cases

### PR DESCRIPTION
**Pull request by kwatters**

### CODE REVIEW

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java
Comment (kyle-kmw) [inline lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java:390]: If there are multiple connections and closing the first one causes an exception, we won't close the other connections because of the throw here.

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java
Comment (kyle-kmw) [inline lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java:302]: If executeUpdate throws an exception the statement won't be closed. The statement can be put inside a try-with-resources.

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java
Comment (kyle-kmw) [inline lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java:180]: I realize this isn't new code but it feels like there should be a TODO/task to clean up the way that the result sets are closed so that they are performed in a finally block or opened in a try-with-resources.

File: lucille-core/src/main/java/com/kmwllc/lucille/core/HashMapDocument.java
Comment (rseitz) [inline lucille-core/src/main/java/com/kmwllc/lucille/core/HashMapDocument.java:616]: new method on Document here -- should add coverage for this in DocumentTest

File: lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java
Comment (rseitz) [inline lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java:233]: we should still check that the connection itself is closed

File: lucille-core/src/main/java/com/kmwllc/lucille/stage/CollapseChildrenDocuments.java
Comment (rseitz) [inline lucille-core/src/main/java/com/kmwllc/lucille/stage/CollapseChildrenDocuments.java:30]: this is treating all child field values as strings, what if some are ints etc?

Comment (kwatters) [reply lucille-core/src/main/java/com/kmwllc/lucille/stage/CollapseChildrenDocuments.java:30]: we would need a plain untyped method on document that returns Object to do this field value copy in a way that worked for all types of objects.  For now, we've chosen string because it generally works with numbers and text...  other field values like dates or bytes probably won't get collapsed down properly here.

Comment (rseitz) [reply lucille-core/src/main/java/com/kmwllc/lucille/stage/CollapseChildrenDocuments.java:39]: you can use the following method on Document:
  public void setOrAdd(String name, Document other)

It takes a field name and another document. It'll add the field from the other document to the current one in a way that preserves the type.

We faced the same problem when implementing the collapsing feature on Connector -- this method, along with   public void setOrAddAll(Document other), was created to handle this situation properly.

File: lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java
Comment (kwatters) [reply lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java:233]: connector.getConnection() no longer exists...  I'd have to add some getters to the class to get to the underlying list connections now.   I feel like isClosed() on the connector is probably ok.

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java
Comment (kwatters) [reply lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java]: no double logging anymore.

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java
Comment (kwatters) [reply lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java]: now we catch it and continue to try to close the rest.

File: lucille-core/src/main/java/com/kmwllc/lucille/stage/CollapseChildrenDocuments.java
Comment (kwatters) [reply lucille-core/src/main/java/com/kmwllc/lucille/stage/CollapseChildrenDocuments.java]: updated to use setOrAdd(fieldname, Document)

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java
Comment (kwatters) [reply lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java]: TODO added.

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java
Comment (kwatters) [reply lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java]: it's in a try with resources now.

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java
Comment (kwatters) [reply lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java]: no longer double logging the exception.

File: lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java
Comment (kyle-kmw) [reply lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java:233]: I think it's good not to have a getter that exposes the underlying connections. I think it would be better to have a test that checks that the connections are really closed though.

I think we can use the connection that dbHelper has open to inspect the number of sessions against the h2 database.

```
  public int checkNumConnections() throws SQLException {
    try(ResultSet rs = RunScript.execute(conn, new StringReader("select count(*) from information_schema.sessions;"))) {
      rs.next();
      return rs.getInt(1);
    }
  }
```
Should be 1 after connector.close() is called (1 being the connection owned by dbHelper)

Comment (kwatters) [reply lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java:233]: cool...  turns out DBTestHelper holds onto a connection from the beginning to the end of the unit tests.  So, I updated the tests to assert the number of connection is 1 before and after each of the test cases.

**Review by kyle-kmw — APPROVED** [review]

File: lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java
Comment (rseitz) [inline lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java:92]: Should call assertEquals(2, dbHelper.checkNumConnections()) or something similar, right before connector close. Otherwise wise we're asserting 1 connection at the beginning and 1 at the end but never seeing it change anywhere.

**Review by rseitz — APPROVED** [review]

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java
Comment (rseitz) [inline lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java:104]: Is there somewhere we know or suspect we're not cleaning up connections?

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java
Comment (rseitz) [inline lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java:167]: is this comment still current?

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java
Comment (rseitz) [inline lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java:226]: so the child IDs are just integers without any prefix based on the parent ID?

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java
Comment (kwatters) [reply lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java:167]: nope , i've removed the comment.

File: lucille-core/src/test/java/com/kmwllc/lucille/stage/CollapseChildrenDocumentsTest.java
Comment (rseitz) [inline lucille-core/src/test/java/com/kmwllc/lucille/stage/CollapseChildrenDocumentsTest.java:39]: should ideally also have a test where the config has dropChildren: false 
and we make sure the children aren't dropped

File: lucille-core/src/test/java/com/kmwllc/lucille/stage/CollapseChildrenDocumentsTest.java
Comment (rseitz) [inline lucille-core/src/test/java/com/kmwllc/lucille/stage/CollapseChildrenDocumentsTest.java:44]: should ideally look at values in childfield1 and childfield2, make sure values are correct and occur in proper order. suggest adding foo1 and bar1 to first child doc and foo2 and bar2 to second child doc so we can test that the ordering is correct after collapse

File: lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java
Comment (rseitz) [inline lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java:179]: outdated TODO?

File: lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java
Comment (rseitz) [inline lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java:182]: outdated TODO?

File: lucille-examples/lucille-joining-db-connector-example/scripts/run_ingest.sh
Comment (rseitz) [inline lucille-examples/lucille-joining-db-connector-example/scripts/run_ingest.sh:2]: where is conf/joining-db-connector-example.conf in this PR?

File: lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java
Comment (rseitz) [inline lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java:1]: looking at this again, I'm seeing a lot of new coverage added in the PR. So how are we specifically flexing the new functionality that's been added to the connector?

File: lucille-core/src/test/java/com/kmwllc/lucille/stage/CollapseChildrenDocumentsTest.java
Comment (kwatters) [reply lucille-core/src/test/java/com/kmwllc/lucille/stage/CollapseChildrenDocumentsTest.java:39]: i added another test to assert that children documents were not dropped.

File: lucille-core/src/test/java/com/kmwllc/lucille/stage/CollapseChildrenDocumentsTest.java
Comment (kwatters) [reply lucille-core/src/test/java/com/kmwllc/lucille/stage/CollapseChildrenDocumentsTest.java:44]: updated unit test to validate the order of the values.

File: lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java
Comment (kwatters) [reply lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java:1]: This changes some details of how the connections are made to the database, and how the result sets are iterated, the existing unit tests covered the functionality and verified that we didn't break anything.

File: lucille-examples/lucille-joining-db-connector-example/scripts/run_ingest.sh
Comment (kwatters) [reply lucille-examples/lucille-joining-db-connector-example/scripts/run_ingest.sh:2]: missing...  I'll add an example.

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java
Comment (rseitz) [reply lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java:107]: We should go through all TODOs in the codebase, but since that's not part of a regular process yet, how about making a ticket for this -- or fixing it with this PR?

**Review by rseitz — APPROVED** [review]
Approving this now to not block merging while I'm away.
Requesting the missing conf file & either a fix to the connection cleanup TODO or a ticket created for it before merging.

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java
Comment (meghanboyd) [reply lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java:107]: Created a ticket LC-316 to address connection cleanup so that the merging of this PR is not otherwise blocked.

**Pull request merged**
